File: Sources/UITextView/UITextView+AutoLayoutFitSize.m
//
//  UITextView+AutoLayoutFitSize.m
//  SmartCity
//
//  Created by ZCGC on 2017/5/13.
//  Copyright © 2017年 NRH. All rights reserved.
//

#import "UITextView+AutoLayoutFitSize.h"
#import "UIView+LayoutConstraints.h"

@implementation UITextView (AutoLayoutFitSize)


- (CGFloat)fitWidth {
    return [self fitWidthOffset:0];
}

- (CGFloat)fitWidthOffset:(CGFloat)os {
    NSLayoutConstraint *cons = self.widthConstraint;
    if (!cons) {
        return 0;
    }
    return cons.constant = [self sizeThatFits:CGSizeMake(999, self.font.pointSize*1.5)].width + os;
}

- (CGFloat)fitHeightForWidth:(CGFloat)width {
    return [self fitHeightForWidth:width offset:0];
}

- (CGFloat)fitHeightForWidth:(CGFloat)width offset:(CGFloat)os {
    NSLayoutConstraint *cons = self.heightConstraint;
    if (!cons) {
        return 0;
    }
    return cons.constant = [self sizeThatFits:CGSizeMake(width, 999)].height + os;
}


@end

File: Sources/UITextView/UITextView+AutoLayoutFitSize.swift
//
//  UITextView+AutoLayoutFitSize.swift
//  SPUIKitCategory module
//
//  Created by LSP on 2020/12/22.
//  Copyright © 2020 LSP. All rights reserved.
//

#if !os(macOS)
import UIKit

/**
 *  使用此分类的API, 需要预先设置宽度约束或高度约束.
 *  宽度/高度约束在 "UIView+Constraints.h" 中定义.
 *  设置约束时, 可在XIB或StoryBoard中关联, 也可以
 *  用代码初始化, 调用Setter即可.
 */
extension UITextView {

    /**
     *  根据文本内容计算宽度并用约束来控制frame.(单行)
     *  @param os: [文本所占的宽度 + os] 即为frame的宽度. 范围:[FLT_MIN ~ FLT_MAX]
     *  @return 计算后的宽度值
     */
    public
    func fitWidth(offset: CGFloat = 0) -> CGFloat {
        if let cons = self.widthConstraint {
            cons.constant = self.sizeThatFits(CGSize(width: 99999, height: self.font?.pointSize ?? 15 * 1.5)).width + offset
            return cons.constant
        }
        return 0
    }

    /**
     *  在指定的宽度之内, 计算文本所占的高度.(单行/多行)
     *  @param width: frame的最大宽度.
     *  @param os: 文本所占的高度 + os 即为frame的高度. 范围:[FLT_MIN ~ FLT_MAX]
     *  @return 计算后的高度值
     */
    public
    func fitHeightFor(width: CGFloat, offset: CGFloat = 0) -> CGFloat {
        if let cons = self.heightConstraint {
            cons.constant = self.sizeThatFits(CGSize(width: width, height: 999999)).height + offset
            return cons.constant
        }
        return 0
    }
}
#endif
